### PR TITLE
[SPARK-42219][CORE] Introducing a config to close all active SparkContexts after the Main method has finished

### PR DESCRIPTION
**Pull request by attilapiros**
### What changes were proposed in this pull request?

Introducing a config to close all active SparkContexts after the Main method has finished.

### Why are the changes needed?

We run into errors after upgrading from Spark 3.1 to Spark 3.2 as the SparkContext get closed right after the starting of the application. It turned out the root cause is [SPARK-34674](https://issues.apache.org/jira/browse/SPARK-34674) which introduced the closing of the SparkContexts after the Main method has finished. For details see https://github.com/apache/spark/pull/32283.

This application was a spark job server built on top of springboot so all the job submits were outside of the main method.

### Does this PR introduce _any_ user-facing change?

With the current default (true) so it is the same behaviour as for YARN.

### How was this patch tested?

Manually.

### CODE REVIEW

**Comment by attilapiros** [comment]
cc @dongjoon-hyun, @HyukjinKwon

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1008]: Shall we put this at the end of this condition?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1008]: If we put this before `args.master.startsWith("k8s")`, it looks like a general configuration which can be used outside `k8s`.

Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1008]: This is a boolean which is easier to evaluate than `String#startsWith` and if it is false the evaluation stops here.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2025]: `3.5.0`? Also, cc @xinrong-meng because she is the release manager of Apache Spark 3.4.0.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2024]: This looks exceptional. Although I understand the situation, it's really weird to have `spark.kubernetes.` configuration in this file.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1008]: This is not a CPU-intensive code part and it doesn't save many CPU cycle. Logical readability is better in this case, @attilapiros .

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:2024]: I'd recommend to move this to K8s module and use config string name instead in `SparkSubmit.scala`.

Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:2024]: Do you mean having `kubernetes` within this file in the core module?

It follows the pattern of:
- spark.yarn.stagingDir
- spark.yarn.isPython
- spark.yarn.kerberos.renewal.excludeHadoopFileSystems
- spark.yarn.dist.pyFiles

Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:2024]: And there is:
- spark.standalone.submit.waitAppCompletion

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:1008]: That's true, I move this at the end of the condition.

**Review by dongjoon-hyun — DISMISSED** [review]
That's a bad example from ancient YARN age, @attilapiros . I'd not re-enforce those bad habits. Instead, I can give you the counter examples like

https://github.com/apache/spark/blob/aeb2a13471e99460e45dbf1a2cf7cc5cc9b74dbc/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala#L308

https://github.com/apache/spark/blob/aeb2a13471e99460e45dbf1a2cf7cc5cc9b74dbc/core/src/main/scala/org/apache/spark/deploy/PythonRunner.scala#L88-L90

**Comment by attilapiros** [comment]
@dongjoon-hyun I have moved the config into the k8s module

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:320]: This PR introduces a new config namespace, `spark.kubernetes.submit.*`. Instead of this naming, shall we use the existing `spark.kubernetes.driver.`? If you want to include `submit` somewhere, you can add it to the last part of the config name.

Comment (attilapiros) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:320]: This config really used at the spark submit so I think it belongs to `spark.kubernetes.submit`. 

If something else will come up in the future which also has effect on the spark submit and Kubernetes related then using the same reasoning we will never start that namespace. I mean there is always a first one.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala
Comment (khalidmammadov) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:322]: What about phrasing it as: `When set to true, Spark on Kubernetes will stop all active Spark contexts once non-shell application' main methods are finished` instead?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:320]: Okay, I'll leave this to you and other committers.

**Comment by attilapiros** [comment]
cc @holdenk

**Comment by attilapiros** [comment]
This pyspark test failure is unrelated:
```
Error:  running /__w/spark/spark/python/run-tests --modules=pyspark-core,pyspark-streaming,pyspark-ml --parallelism=1 ; received return code 255
Error: Process completed with exit code 19.
```

**Comment by pan3793** [comment]
SPARK-42698(https://github.com/apache/spark/pull/40314) is aiming to expand the scope of stopping SparkContext after `runMain` exiting to all cluster managers, if it is accepted, the configuration proposed by this PR should expand scope as well.

**Review by srowen — COMMENTED** [review]
Should this be specific to Kubernetes?
Does it need to be a config or a method you can call?
Actually, why would you not kill the contexts after main exits in any case?

**Comment by attilapiros** [comment]
@srowen 

> Should this be specific to Kubernetes? 

The original https://github.com/apache/spark/pull/32283 was Kubernetes specific. This PR just adds a new config to have the old behaviour as default but make the new one also available. 

> Does it need to be a config or a method you can call? 

Unfortunately there is a use case for both behaviour. See the next point.

> Actually, why would you not kill the contexts after main exits in any case?

I bumped into this change when I analysed an application where spark was used as a job server. 
With the older spark it was running just fine but after an update it was stopping in the very beginning.
The app was built on top of springboot where job requests was served via REST.     
In a springboot app the main method just initialises / registers the REST handlers and the serving of the new requests done on separate threads. With the new behaviour the Spark context was closed right after the initialisation.

**Comment by attilapiros** [comment]
cc @mridulm

**Comment by zhouyifan279** [comment]
Our customer also encountered this issue recently. They are migrating their spark job server (a spring boot application) from Spark 2.4 to Spark 3.
They feel surprised that only Spark 3.1.2 stops SparkContext after job server started.

**Comment by attilapiros** [comment]
I am closing this PR. Job servers has the option to do blocking call in the main method to avoid the auto stopping of the active spark contexts.

**Pull request closed without merging**
